File: ws/first_ws/src/service_pkg/src/client_custom_srv.cpp
//
// Created by zcb on 2021/4/9.
//

#include <ros/ros.h>
#include <iostream>
#include <demo_srvs/NumOption.h>

using namespace std;

int main(int argc, char **argv) {

    string nodeName = "cpp_client_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    string serviceName = "/cpp/service";
    ros::ServiceClient client = node.serviceClient<demo_srvs::NumOption>(serviceName);

    client.waitForExistence(); // 如果client先于 server 起来，可以一直等server

    demo_srvs::NumOption service;
    service.request.a = 10;
    service.request.b = 5;
    service.request.option = "*";
    bool flag = client.call(service);
    if (flag) {
        ROS_INFO_STREAM("result is: " << service.response.result);
    } else {
        ROS_DEBUG_STREAM("call fail");
    }


    return 0;
}

File: ros/ws01/first_ws/src/service_pkg/scripts/client_custom_srv.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_srvs.srv import NumOption, NumOptionRequest, NumOptionResponse

if __name__ == '__main__':
    nodeName = "python_client_node"
    rospy.init_node(nodeName)

    serviceName = "/python/service"

    # 与c++ api 不一致，
    client = rospy.ServiceProxy(serviceName, NumOption)
    client.wait_for_service()

    # 与c++ api 不一致,这里只传 request
    request = NumOptionRequest()
    request.a = 10
    request.b = 20
    request.option = "%"

    # 与c++ api 又不一致，艹艹艹艹艹艹
    try:
        response = client.call(request)  # response's type is AddTwoIntsResponse
        if isinstance(response, NumOptionResponse):
            rospy.loginfo("result is %d" % response.result)
    except Exception:
        print("call fail")
